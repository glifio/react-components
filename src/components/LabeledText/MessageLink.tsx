import PropTypes from 'prop-types'
import { useMemo } from 'react'
import { useEnvironment } from '../../services/EnvironmentProvider'

import truncateAddress from '../../utils/truncateAddress'
import {
  LabeledLink,
  LabeledLinkProps,
  LabeledLinkPropTypes
} from './LabeledLink'

export const MessageLink = ({
  cid,
  shouldTruncate,
  ...labeledLinkProps
}: MessageLinkProps) => {
  const { explorerUrl } = useEnvironment()
  const linkText = useMemo(
    () => (shouldTruncate ? truncateAddress(cid) : cid),
    [cid, shouldTruncate]
  )
  const href = `${explorerUrl}/message/?cid=${cid}`
  return (
    <LabeledLink
      href={href}
      linkText={linkText}
      copyText={cid}
      {...labeledLinkProps}
    />
  )
}

export type MessageLinkProps = {
  cid?: string
  shouldTruncate?: boolean
} & Omit<LabeledLinkProps, 'href' | 'linkText' | 'copyText'>

const { href, linkText, copyText, ...messageLinkPropTypes } =
  LabeledLinkPropTypes

MessageLink.propTypes = {
  cid: PropTypes.string,
  shouldTruncate: PropTypes.bool,
  ...messageLinkPropTypes
}

MessageLink.defaultProps = {
  shouldTruncate: true
}
