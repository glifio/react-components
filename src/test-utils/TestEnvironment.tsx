import { ReactNode } from 'react'
import { MockedProvider, MockedResponse } from '@apollo/client/testing'
import { CoinType } from '@glif/filecoin-address'
import LotusRPCEngine from '@glif/filecoin-rpc-client'
import { Environment, Network, networks } from '../services/EnvironmentProvider'

const OptionalApolloMockProvider = ({ children, mocks, withApollo }) => {
  if (withApollo || mocks.length > 0)
    return <MockedProvider mocks={mocks}>{children}</MockedProvider>

  return <>{children}</>
}

export const TestEnvironment = ({
  children,
  withApollo,
  apolloMocks
}: TestEnvironmentProps) => (
  <Environment
    coinType={CoinType.TEST}
    networkName={Network.CALIBRATION}
    nodeStatusUrl={networks[Network.CALIBRATION].nodeStatusUrl}
    graphUrl={networks[Network.CALIBRATION].graphUrl}
    graphSecure={networks[Network.CALIBRATION].graphSecure}
    lotusApiUrl={networks[Network.CALIBRATION].lotusApiUrl}
    explorerUrl='https://explorer-calibration.glif.link'
    lotusApi={
      new LotusRPCEngine({
        apiAddress: networks[Network.CALIBRATION].lotusApiUrl
      })
    }
  >
    <OptionalApolloMockProvider withApollo={withApollo} mocks={apolloMocks}>
      {children}
    </OptionalApolloMockProvider>
  </Environment>
)

type TestEnvironmentProps = {
  children: ReactNode
  withApollo?: boolean
  apolloMocks?: MockedResponse[]
}

TestEnvironment.defaultProps = {
  withApollo: false,
  apolloMocks: []
}
