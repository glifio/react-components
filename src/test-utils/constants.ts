export const WALLET_ADDRESS = 't1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi'
export const WALLET_ID = 't0101'

export const WALLET_ADDRESS_2 = 't1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a'
export const WALLET_ID_2 = 't0102'

export const WALLET_ADDRESS_3 = 't1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq'
export const WALLET_ID_3 = 't0103'
