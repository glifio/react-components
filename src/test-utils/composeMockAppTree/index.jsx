/* eslint-disable react/prop-types */
import React from 'react'
import { mockWalletProviderInstance } from '../../../__mocks__/@glif/filecoin-wallet-provider'
import WalletProviderWrapper from '../../services/WalletProvider'
import { initialState as walletProviderInitialState } from '../../services/WalletProvider/state'
import { composeWalletProviderState } from './composeState'
import { TestEnvironment } from '../TestEnvironment'

jest.mock('../../services/WalletProvider')

const Index = (statePreset = 'preOnboard', options = {}) => {
  // here you can pass a walletProviderInitialState and a preset to shape the store how you want it for testing
  const initialState = composeWalletProviderState(
    options?.walletProviderInitialState || walletProviderInitialState,
    statePreset
  )

  let walletProviderCache = { ...initialState }

  const cacheWalletProviderState = state => {
    walletProviderCache = { ...state }
    return <></>
  }

  const getWalletProviderState = () => walletProviderCache

  const Tree = ({ children }) => {
    return (
      <TestEnvironment>
        <WalletProviderWrapper
          options={options}
          statePreset={statePreset}
          getState={cacheWalletProviderState}
          initialState={initialState}
        >
          {children}
        </WalletProviderWrapper>
      </TestEnvironment>
    )
  }

  return {
    Tree,
    walletProvider: mockWalletProviderInstance,
    getWalletProviderState
  }
}

/**
 * This function is a wrapper that mocks everything the filecoin app needs for testing
 *
 * VALID OPTIONS:
 * mockConverterInstance (an object that stubs the converter functionality)
 * walletProviderReducer (a reducer to stub the state handling of the wallet provider)
 * walletProviderInitialState(a reducer to stub the state of the walletProvider context)
 * walletProviderDispatch(a dispatcher to mock/read calls to the walletProvider's state)
 * pathname
 * query
 */

export default Index
