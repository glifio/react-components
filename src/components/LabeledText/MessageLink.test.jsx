import { render } from '@testing-library/react'
import { MessageLink } from './MessageLink'
import { TestEnvironment } from '../../test-utils/TestEnvironment'

describe('MessageLink', () => {
  test('it renders correctly', () => {
    const { container } = render(
      <TestEnvironment>
        <MessageLink
          label='Message'
          cid='bafy2bzacebgtugc5fzoyyit3cis4emhuapnbq43lttiubvh6ubn5mt7o7eja6'
        />
      </TestEnvironment>
    )

    expect(container.firstChild).toMatchSnapshot()
  })
})
