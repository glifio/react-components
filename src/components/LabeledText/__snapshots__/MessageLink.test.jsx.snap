// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageLink it renders correctly 1`] = `
.c0 h4 {
  margin: 0;
  color: var(--gray-dark);
}

.c0 p {
  margin: 0;
  line-height: 1.5;
}

.c3 {
  width: 24;
  height: 24;
}

.c2 {
  position: relative;
  cursor: pointer;
}

.c2 svg {
  -webkit-transition: -webkit-transform 0.1s ease-out;
  -webkit-transition: transform 0.1s ease-out;
  transition: transform 0.1s ease-out;
}

.c2 .copy-text {
  position: absolute;
  top: calc(50% - 0.2em);
  left: calc(100% + 0.25em);
  color: var(--purple-medium);
  opacity: 0;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: opacity 0.2s ease-in;
  transition: opacity 0.2s ease-in;
  pointer-events: none;
}

.c2:hover svg {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
}

.c2:hover .copy-text {
  opacity: 1;
  pointer-events: all;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.25em;
  line-height: 1.5;
}

.c1 > a {
  color: var(--purple-medium);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 > a:hover {
  color: var(--purple-medium);
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1 > a svg {
  -webkit-transition: 0.24s ease-in-out;
  transition: 0.24s ease-in-out;
  vertical-align: middle;
}

.c1 > a svg:hover {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
  fill: var(--purple-medium);
}

<div
  class="c0"
>
  <h4>
    Message
  </h4>
  <div
    class="c1"
    color="var(--purple-medium)"
  >
    <a
      href="https://explorer-calibration.glif.link/message/?cid=bafy2bzacebgtugc5fzoyyit3cis4emhuapnbq43lttiubvh6ubn5mt7o7eja6"
      rel="noreferrer noopener"
      target="_blank"
    >
      bafy2b ... o7eja6
    </a>
    <span
      class="c2"
      color="var(--purple-medium)"
      role="button"
    >
      <svg
        class="c3"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="#1A0066"
          height="12"
          width="12"
          x="8"
          y="4"
        />
        <rect
          fill="#E0D7FE"
          height="12"
          width="12"
          x="4"
          y="8"
        />
      </svg>
    </span>
  </div>
</div>
`;
